### PR TITLE
feat: add missing zks namespace methods to zksync provider

### PR DESCRIPTION
**Pull request by vasyl-ivanchuk**
Add the following zks namespace methods to zksync provider:
* `zks_estimateGasL1ToL2`
* `zks_getBridgehubContract`
* `zks_getBridgeContracts`
* `zks_getBaseTokenL1Address`
* `zks_getAllAccountBalances`
* `zks_getL2ToL1MsgProof`
* `zks_getL2ToL1LogProof`
* `zks_getBlockDetails`
* `zks_getTransactionDetails`
* `zks_getRawBlockTransactions`
* `zks_getL1BatchDetails`
* `zks_getBytecodeByHash`
* `zks_getL1BatchBlockRange`
* `zks_getL1GasPrice`
* `zks_getFeeParams`
* `zks_getProtocolVersion`
* `zks_getProof`

### CODE REVIEW

File: src/provider/mod.rs [hunk 1]
```diff
@@ -31,6 +34,268 @@ pub struct Eip712Fee {
     pub max_priority_fee_per_gas: u128,
 }
 
+/// Response type for `zks_getBridgeContracts`.
```
Comment (popzxc) [inline src/provider/mod.rs:37]: I guess it's time to create `types` module so that types are stored separately.

File: src/provider/mod.rs [hunk 2]
```diff
@@ -59,13 +324,193 @@ where
...
+    /// Lists confirmed tokens.
+    /// Confirmed in the method name means any token bridged to ZKsync Era via the official bridge.
+    ///
+    /// The tokens are returned in alphabetical order by their symbol.
+    /// This means the token id is its position in an alphabetically sorted array of tokens.
+    fn get_confirmed_tokens(&self, from: u32, limit: u8) -> ProviderCall<T, (u32, u8), Vec<Token>> {
+        self.client()
+            .request("zks_getConfirmedTokens", (from, limit))
+            .into()
+    }
```
Comment (popzxc) [inline src/provider/mod.rs:373]: This method is deprecated and will be removed in the future, no need to add it.

File: src/provider/mod.rs [hunk 3]
```diff
@@ -59,13 +324,193 @@ where
...
+    fn get_transaction_details(
+        &self,
+        tx_hash: B256,
+    ) -> ProviderCall<T, (B256,), Option<TransactionDetails>> {
+        self.client()
+            .request("zks_getTransactionDetails", (tx_hash,))
+            .into()
+    }
+
+    /// Lists transactions in a block without processing them.
```
Comment (popzxc) [inline src/provider/mod.rs:432]:
```suggestion
    /// Lists transactions in a native encoding (e.g. that has more details, but does not
    /// adhere to the "common" Web3 Transaction interface).
```

File: src/provider/mod.rs [hunk 4]
```diff
@@ -31,6 +34,268 @@ pub struct Eip712Fee {
...
+    pub signature: Bytes,
+    pub transaction_type: String,
+    pub input: Option<InputData>,
+    pub paymaster_params: PaymasterParams,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub enum ExecuteTransactionCommon {
+    L2(L2TxCommonData),
+}
```
Comment (popzxc) [inline src/provider/mod.rs:289]: I believe L1 and protocol upgrade transactions can also be queried through API.

File: src/provider/mod.rs [hunk 5]
```diff
@@ -31,6 +34,268 @@ pub struct Eip712Fee {
...
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum BlockStatus {
+    Sealed,
+    Verified,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct BlockDetailsBase {
```
Comment (popzxc) [inline src/provider/mod.rs:66]: Given that it's used in a single place via `serde(flatten)`, maybe it makes sense to move these fields to `BlockDetails` directly, or do you anticipate other use cases for this structure?

File: src/provider/mod.rs [hunk 6]
```diff
@@ -59,13 +324,193 @@ where
...
+    /// The API has a similar behaviour to eth_sendRawTransaction
+    /// but with some extra data returned from it.
+    fn send_raw_transaction_with_detailed_output(
+        &self,
+        tx_bytes: Bytes,
+    ) -> ProviderCall<T, (Bytes,), TransactionDetailedResult> {
+        self.client()
+            .request("zks_sendRawTransactionWithDetailedOutput", (tx_bytes,))
+            .into()
+    }
```
Comment (popzxc) [inline src/provider/mod.rs:513]: I can anticipate that this method will be deprecated in the future, so let's not add it for now as well.

**Review by popzxc — COMMENTED** [review]
Mostly LGTM, well done!

File: src/provider/mod.rs (hunk 2 above)
Comment (vasyl-ivanchuk) [reply src/provider/mod.rs]: Deleted.

File: src/provider/mod.rs (hunk 3 above)
Comment (vasyl-ivanchuk) [reply src/provider/mod.rs]: Applied.

File: src/provider/mod.rs (hunk 4 above)
Comment (vasyl-ivanchuk) [reply src/provider/mod.rs]: Good catch, I've added both.

File: src/provider/mod.rs (hunk 5 above)
Comment (vasyl-ivanchuk) [reply src/provider/mod.rs]: `BlockDetails` was used in two places, but as we discussed such approach hurts the usage experience, so I removed `BlockDetails` and added corresponding fields to the two structs that were using it.

File: src/provider/mod.rs (hunk 6 above)
Comment (vasyl-ivanchuk) [reply src/provider/mod.rs]: Got it, removed.

**Review by popzxc — APPROVED** [review]

**Pull request merged**